aten/src/ATen/native/Sorting.cpp

```diff
@@ -94,6 +91,169 @@ void quick_select_template(
   } while (true);
 }
 
+// Returns pointer to pivot used to partition range [first, last]
+// There should be at least 3 elements in [first, last]
+template <typename random_it, typename comp>
+random_it partition(random_it first, random_it last, comp lt) {
+  // Use median of three for pivot choice
+  random_it pivot = first + (std::distance(first, last) >> 1);
+  std::swap(first[1], *pivot);
+  if (lt(*last, first[1])) {
+    std::swap(first[1], *last);
+  }
+  if (lt(*last, *first)) {
+    std::swap(*first, *last);
+  }
+  if (lt(*first, first[1])) {
+    std::swap(first[1], *first);
+  }
+  pivot = first++;
+
+  do {
+    do {
+      ++first;
+    } while (lt(*first, *pivot));
+    do {
+      --last;
+    } while (lt(*pivot, *last));
+    if (last < first) {
+      break;
+    }
+    std::swap(*first, *last);
+  } while (true);
+  std::swap(*pivot, *last);
+
+  return last;
+}
+
+template <typename random_it, typename kth_iterator, typename comp>
+void quickselect_helper(
+    random_it start,
+    random_it first,
+    random_it last,
+    kth_iterator kbegin,
+    kth_iterator kend,
+    comp lt) {
+  // Base case: one element or no kth values
+  if (last <= first || kbegin == kend) {
+    return;
+  }
+
+  // Base case: two elements
+  if (last == first + 1) {
+    if (lt(*last, *first)) {
+      std::swap(*first, *last);
+    }
+    return;
+  }
+
+  random_it pivot = partition(first, last, lt);
+  int64_t pivot_rank = std::distance(start, pivot);
+
+  auto lt_kend = std::lower_bound(kbegin, kend, pivot_rank);
+  quickselect_helper(start, first, pivot - 1, kbegin, lt_kend, lt);
+
+  auto gt_kbegin = std::upper_bound(lt_kend, kend, pivot_rank);
+  quickselect_helper(start, pivot + 1, last, gt_kbegin, kend, lt);
+}
+
+// Rearranges elements in range [first, last) such that kths elements
+// are in their sorted location.
+template <typename random_it, typename comp>
+void quickselect(
+    random_it first,
+    random_it last,
+    std::vector<int64_t> kths,
+    comp lt) {
+  return quickselect_helper(
+      first, first, last - 1, kths.begin(), kths.end(), lt);
+}
+
+void quantile_cpu_impl(
+    Tensor& out,
+    const Tensor& self,
+    const Tensor& q,
+    const optional<int64_t>& _dim,
+    bool ignore_nan) {
+  int64_t dim = at::maybe_wrap_dim(_dim.value_or(0), self.dim(), true);
+
+  // Checks that all q values are between 0 and 1, inclusive
+  // NOTE: this check is only performed when running on the CPU to avoid
+  // synchronizing an accelerator with the CPU
+  TORCH_CHECK(
+      q.ge(0).logical_and_(q.le(1)).all().item<bool>(),
+      "quantile() q values must be in the range [0, 1]");
+
+  // If dim is None flatten. Move dimension to reduce as last dim and flatten
```

Comment:
This could use a slight tweak. Something like... "If dim is None then quantiles are computed on the flattened tensor. If dim is not None, ..."?